usr/lib/cca_stdll/cca_specific.c

```diff
@@ -2977,6 +2977,90 @@ static CK_BBOOL ccatok_pkey_option_disabled(STDLL_TokData_t *tokdata)
     return CK_FALSE;
 }
 
+/*
+ * This function is called whenever a new object is created. It currently sets
+ * attribute CKA_IBM_PROTKEY_EXTRACTABLE according to the PKEY_MODE token
+ * option, but may also be used for other token options and attrs in future.
+ */
+CK_RV token_specific_set_attrs_for_new_object(STDLL_TokData_t *tokdata,
+                                              CK_OBJECT_CLASS class,
+                                              CK_ULONG mode, TEMPLATE *tmpl)
+{
+    struct cca_private_data *cca_data = tokdata->private_data;
+    CK_ATTRIBUTE *pkey_attr = NULL, *ecp_attr = NULL;
+    CK_BBOOL add_pkey_extractable = CK_FALSE;
+    CK_BBOOL btrue = CK_TRUE;
+    CK_RV ret;
+
+    UNUSED(mode);
+
+    if (class != CKO_SECRET_KEY && class != CKO_PRIVATE_KEY &&
+        class != CKO_PUBLIC_KEY)
+        return CKR_OK;
+
+    switch (cca_data->pkey_mode) {
+    case PKEY_MODE_DISABLED:
+        /* Nothing to do */
+        break;
+    case PKEY_MODE_DEFAULT:
+        /*
+         * If the application did not specify pkey-extractable, all keys get
+         * pkey-extractable=false. This was already set by default, so
+         * nothing to do here.
+         */
+        break;
+    case PKEY_MODE_ENABLED:
+        /*
+         * All secret/private keys and all EC public keys where CPACF supports
+         * the related curve, get pkey-extractable=true
+         */
+        switch (class) {
+        case CKO_PUBLIC_KEY:
+            if (template_attribute_get_non_empty(tmpl, CKA_EC_PARAMS, &ecp_attr) == CKR_OK &&
+                pkey_op_supported_by_cpacf(cca_data->msa_level, CKM_ECDSA, tmpl))
+                add_pkey_extractable = CK_TRUE;
+                /*
+                 * Note that the explicit parm CKM_ECDSA just tells the
+                 * function that it's not AES here. It covers all EC mechs
+                 */
+            break;
+        default:
+            add_pkey_extractable = CK_TRUE;
+            break;
+        }
+
+        if (add_pkey_extractable) {
+            if (!template_attribute_find(tmpl, CKA_IBM_PROTKEY_EXTRACTABLE, &pkey_attr)) {
```

Comment:
Use template_attribute_get_bool

Reply:
Well, might be ok to use template_attribute_find here, since its just about checking the existence of the attribute here.

Reply:
Done. We may want to fix token_specific_set_attrs_for_new_object in ep11_specific.c in the same way.

Reply:
Hm, I think ists better the way it was before.... I don't like the local variable `CK_BBOOL pkey_attr_value` within the block, and also the check `!= CKR_OK || pkey_attr_value == CK_FALSE)` seems wrong. Why checking for the variable being FALSE ?
Lets revert this change to how it was before.

Reply:
Done.